sound/soc/sof/ipc4.c

```diff
@@ -108,9 +108,9 @@ static int ipc4_tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *ms
 			msg->header, msg->extension);
 		return -ETIMEDOUT;
 	} else if (msg->reply_error) {
-		dev_err(sdev->dev, "ipc error for msg 0x%x : 0x%x\n",
-			msg->header, msg->extension);
-		return -EIO;
+		dev_err(sdev->dev, "ipc error for msg: 0x%x extension: 0x%x : %d\n",
+			msg->header, msg->extension, msg->reply_error);
+		return msg->reply_error;
```

Comment:
why would we trust an error code returned in case of an error? how useful would that be anyways since there's no guarantee that the firmware error code matches the Linux ones, so there's a risk of higher levels interpreting the firmware error code as a linux one.

I just don't see the rationale here, and that applies to the initial code in IPCv3 as well. It makes no sense to mix -ETIMEDOUT or a firmware error code in the same routine.

Reply:
Given we use linux error codes I figured that was the model. Hence why we return the codes from the FW. Why would we not want to propogate errors all the way to the top? Removing context and asking errors sounds like an antipattern.

Reply:
IPC4 works a bit different than IPC3 in the sense that an IPC would never return an error ie the immediate reply never contains the actual IPC processing error. It is handled in process_reply just above. So I wonder if we even need this else case. @ujfalusi

Reply:
we just print error in process_reply and more info such as header & extension in tx_wait_done, so I prefer to keep it